metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -354,6 +357,25 @@ public ControllerResult<BrokerRegistrationReply> registerBroker(
             throw new BrokerIdNotRegisteredException("Controller is in pre-migration mode and cannot register KRaft " +
                 "brokers until the metadata migration is complete.");
         }
+
+        if (featureControl.metadataVersion().isDirectoryAssignmentSupported()) {
```

Comment:
If the metadata version does not support directory assignment then it shouldn't be sent by the brokers, right?

And if it does support directory assignment but the brokers are not sending directories, we need to fail the heartbeat.